### DISCUSSION TITLE
Github action defined as submodule in a repo

### DISCUSSION THREAD

**Discussion by PuneethaPai** [community/community#26426] [Actions]
<p><code>jobs.steps.uses</code> can refer to action defined over a public repo, docker, etc.</p>
<p>If action defined over a subdir you can refer to it like: <code>{owner}/{repo}/{path/to/action}@{ref}</code></p>
<p>docs example: <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-using-a-public-action-in-a-subdirectory" rel="noopener nofollow ugc">https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-using-a-public-action-in-a-subdirectory</a></p>
<pre><code class="lang-yaml">jobs:
  my_first_job:
    steps:
      - name: My first step
        uses: actions/aws/ec2@main

</code></pre>
<p>But this fails if <code>path/to/action</code> is actually a git sub-module to the repo defining the action.</p>
<p>I get following error like<br>
<code> Can't find 'action.yml', 'action.yaml' or 'Dockerfile' under '/home/runner/work/_actions/PuneethaPai/checkout/submodule/.github/actions/checkout'. Did you forget to run actions/checkout before running your local action?</code>.</p>
<p>So I think when action is defined in a public repo, git-runner clones it, but doesn’t clone the sub-modules defined in them. Hence I am getting above error.</p>
<p>Workaround is to check-in the whole action code in the path, but it’s not a best way when you have advantages of adding it as sub-module thingy.</p>

**Comment by airtower-luna** [comment]
<p>Couldn’t you refer directly to the submodule repo (or a directory in it) with the <code>uses:</code> keyword? That way the submodule wouldn’t have to be checked out first, or at least not to run the action. :slightly_smiling_face:</p>
<p>The only disadvantage I see is that if the submodule HEAD has to exactly match the action ref you’ll have to update both at the same time.</p>

**Comment by PuneethaPai** [comment]
<p>Thanks <a class="mention" href="/airtower-luna">@airtower-luna</a>.<br>
Actually the real problem is, in our company with Github Enterprise, admins have disabled the use of public actions by restricting to use only local actions, i.e actions which are inside our organization.</p>
<p>More Info:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/d/f/dfa28321983f237d6653c627d289bc9784836246.png" data-download-href="/uploads/short-url/vUmEIAj5GNCbxfIje9yn2xWUL7o.png?dl=1" title="Screenshot 2020-10-29 at 9.05.33 AM" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/21996583/181097611-319b41e7-d36f-4179-ab0a-1b3bd00798e7.png" alt="Screenshot 2020-10-29 at 9.05.33 AM" data-base62-sha1="vUmEIAj5GNCbxfIje9yn2xWUL7o" width="690" height="353" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/d/f/dfa28321983f237d6653c627d289bc9784836246_2_690x353.png, https://aws1.discourse-cdn.com/github/optimized/3X/d/f/dfa28321983f237d6653c627d289bc9784836246_2_1035x529.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/d/f/dfa28321983f237d6653c627d289bc9784836246.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/d/f/dfa28321983f237d6653c627d289bc9784836246_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot 2020-10-29 at 9.05.33 AM</span><span class="informations">1350×692 58.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>To be frank, we can bypass this check by committing whole action code/repo as part of our application repo, but thought sub-module approach would be neater.<br>
Also we can clone required actions repo into our organization, etc. (Which we don’t have enough privilage to do :rofl:  )</p>
<p>PS: I understand that, it is not good to by pass this check in any way unless we get approval from admins :blush: :wink:</p>

**Comment by Simran-B** [comment]
<p>You need to checkout with submodules:</p>
<pre><code class="lang-yaml">    steps:
      - name: Checkout repository
        uses: actions/checkout@v2
        with:
          submodules: recursive
      - name: Use submodule action
        uses: ./.github/actions/checkout
        with:
          ...
</code></pre>
<p>See <a href="https://github.com/actions/checkout" rel="noopener nofollow ugc">https://github.com/actions/checkout</a>:</p>
<blockquote>
<p>Whether to checkout submodules: <code>true</code> to checkout submodules or <code>recursive</code> to recursively checkout submodules. Default: <code>false</code></p>
</blockquote>

**Comment by PuneethaPai** [comment]
<p>Yes <a class="mention" href="/simran-b">@Simran-B</a>.</p>
<p>But <code>uses: actions/checkout@v2</code> itself fails as this is public repo outside organization.<br>
As shared above, our admins have restricted to use only local actions.<br>
Hence I was trying to add this repo: <code>actions/checkout</code> as sub-module to our internal repo.</p>
<p>Let me know if this clarifies.</p>
<p>Thanks</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>I guess in that case you’ll unfortunately have to clone the repository manually, using git commands in a <code>run</code> step.</p>
<p>That said, it seems a questionable choice to use Github for hosting and CI and then not allow the official actions… Maybe you can convince your admins to allow these basic actions, or at least create forks within your organization?</p>

**Comment by npalm** [comment]
<p>We think it would be a good idea when the checkout action supports clonimng private submouldes by providing a different SSH key. The PAT token is not owrking in many cases since it is bounded to a specific user. With a SSH key pair you can clone the repo as submodule by setting up a deploy key. <a href="https://github.com/actions/checkout/issues/287" rel="noopener nofollow ugc">https://github.com/actions/checkout/issues/287</a></p>

**Comment by Simran-B** [comment]
<p>Sorry, I misunderstood the problem. The easiest solution would indeed be to allow Actions by GitHub:</p>
<p><img src="https://user-images.githubusercontent.com/7819991/181097615-ddefa3ae-841a-4b28-b042-f82d4abfcab3.png" alt="image" data-base62-sha1="qra4naed3X8L93YbrgM70a4Cg11" width="325" height="144"></p>
<p>(I wonder if local actions need to be added to the allowlist if <em>Allow select actions</em> is enabled…)</p>
<aside class="quote no-group" data-username="PuneethaPai" data-post="1" data-topic="139204">
<div class="title">
<div class="quote-controls"></div>
 PuneethaPai:</div>
<blockquote>
<p>Workaround is to check-in the whole action code in the path</p>
</blockquote>
</aside>
<p>Actually no. It’s a chicken and egg problem. The runner starts out with a fresh virtual environment. You want to checkout the repo using a local copy of the checkout action, but it doesn’t exist - you first need to checkout the repo… The runner seems to retrieve public actions automatically, so that the workflow run can be bootstrapped by fetching <code>actions/checkout</code> and using it to checkout the repo. Using any means provided by the virtual environment to retrieve at least the checkout action seems to be unavoidable. You could fork the official repo if the code has to be within your own org too.</p>

**Comment by Simran-B** [comment]
<p>Weird, if I fetch the three core files of the checkout action with curl, I can use it as local action to checkout a repo. However, the following errors occur when <code>.temp/checkout</code> is run:</p>
<p>Error: Unable to process command ‘::add-matcher::/home/runner/work/olive-ci/olive-ci/temp/checkout/dist/problem-matcher.json’ successfully.</p>
<blockquote>
<p>Error: Unable to process command ‘::add-matcher::/home/runner/work/test-ci/test-ci/temp/checkout/dist/problem-matcher.json’ successfully.</p>
<p>Error: Could not find a part of the path ‘/home/runner/work/test-ci/test-ci/temp/checkout/dist/problem-matcher.json’.</p>
</blockquote>
<p>With <code>continue-on-error: true</code> it will simply be ignored and the checkout does work! But there is another error in <strong>Post checkout repository</strong> (what the heck?)</p>
<blockquote>
<p>Error: Can’t find ‘action.yml’, ‘action.yaml’ or ‘Dockerfile’ under ‘/home/runner/work/test-ci/test-ci/temp/checkout’. Did you forget to run actions/checkout before running your local action?</p>
</blockquote>
<p>BTW. local actions do not need to be added to the allowlist. And it’s not limited to the very repo, it can be any repo in the same account / organization.</p>

**Comment by PuneethaPai** [comment]
<p>I was thinking, github runners clone actions repo before running them.</p>
<p><code>uses:action/repo</code></p>
<p>Can’t they clone submodules in the repo as well?</p>
<pre><code class="lang-bash">git clone --recurse-submodules &lt;action/repo.git&gt;
</code></pre>
<p>Is that a bad feature request?</p>
<p>Thanks</p>

**Comment by Simran-B** [comment]
<p>Huh. Local action means an action whose repository is in the same user account or org. It’s not restricted to the very repository. That means, you can simply fork <code>actions/checkout</code> and use it in your workflow like <code>uses: {your-org}/checkout@v2</code>. No need to add this action as submodule. You can then checkout the repo with submodules to use any of the actions you added as submodules.</p>

**Comment by PuneethaPai** [comment]
<p>As I said above we are dependent on our admins for this too.</p>
<blockquote>
<blockquote>
<p>Also we can clone required actions repo into our organization, etc. (Which we don’t have enough privilage to do :rofl: )</p>
</blockquote>
</blockquote>

**Comment by Simran-B** [comment]
<p>Ah, I didn’t read that as forking. Well, if they don’t give you enough permissions, how are you supposed to do your job? If they don’t trust you as employee, then they shouldn’t trust GitHub or other cloud services either and go for something on-premise.</p>

**Comment by PuneethaPai** [comment]
<p>Thanks everyone.</p>
<p>I’ll share this thread with admins and hopefully I’ll be able to pursued them to realise the silliness of disabling public action and remove this blocker.</p>
<p>Thanks</p>
